caas/kubernetes/provider/exec/copy.go

```diff
@@ -0,0 +1,220 @@
+	if err := params.validate(); err != nil {
+		return errors.Trace(err)
+	}
+	if params.Src.PodName != "" {
+		return c.copyFromPod(params, cancel)
+	}
+	if params.Dest.PodName != "" {
+		return c.copyToPod(params, cancel)
+	}
+	return nil
+}
+
+func (c client) copyFromPod(params CopyParam, cancel <-chan struct{}) error {
+	// TODO(caas): implement when we need later.
+	return errors.NotSupportedf("copy from pod")
+}
+
+// this is inspired by kubectl cmd package.
+// - https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/cp/cp.go
+func (c client) copyToPod(params CopyParam, cancel <-chan struct{}) (err error) {
+	src := params.Src
+	dest := params.Dest
+	logger.Debugf("coping from %v to %v", src, dest)
+
+	if _, err = os.Stat(src.Path); err != nil {
+		return errors.NewNotValid(nil, fmt.Sprintf("%q does not exist on local", src.Path))
+	}
+
+	if dest.Path != "/" && strings.HasSuffix(dest.Path, "/") {
+		dest.Path = strings.TrimSuffix(dest.Path, "/")
+	}
+
+	if err = c.checkRemotePathIsDir(dest, cancel); err == nil {
+		dest.Path = path.Join(dest.Path, path.Base(src.Path))
+	}
+
+	reader, writer := c.pipGetter()
+
+	go func() {
+		defer writer.Close()
+		err = makeTar(src.Path, dest.Path, writer)
+		if err != nil {
+			logger.Errorf("make tar %q failed: %v", src.Path, err)
+		}
+	}()
+
+	cmds := []string{"tar", "-xmf", "-"}
+	if params.noPreserveOwnershipPermissions {
+		cmds = []string{"tar", "--no-same-permissions", "--no-same-owner", "-xmf", "-"}
+	}
+	destDir := path.Dir(dest.Path)
+	if len(destDir) > 0 {
+		cmds = append(cmds, "-C", destDir)
+	}
+	var stdout, stderr bytes.Buffer
+	execParams := ExecParams{
+		PodName:       dest.PodName,
+		ContainerName: dest.ContainerName,
+		Commands:      cmds,
+		Stdin:         reader,
+		Stdout:        &stdout,
+		Stderr:        &stderr,
+	}
+	return errors.Trace(c.Exec(execParams, cancel))
+}
+
+func (c client) checkRemotePathIsDir(rec FileResource, cancel <-chan struct{}) error {
+	if rec.PodName == "" {
+		return errors.NotValidf("empty pod name")
+	}
+	var stdout, stderr bytes.Buffer
+	execParams := ExecParams{
+		PodName:       rec.PodName,
+		ContainerName: rec.ContainerName,
+		Commands:      []string{"test", "-d", rec.Path},
+		Stdout:        &stdout,
+		Stderr:        &stderr,
+	}
+	return errors.Trace(c.Exec(execParams, cancel))
+}
+
+// Based on code from https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/cp/cp.go
+func makeTar(srcPath, destPath string, writer io.Writer) error {
+	tarWriter := tar.NewWriter(writer)
+	defer tarWriter.Close()
+	srcPath = path.Clean(srcPath)
+	destPath = path.Clean(destPath)
+	return recursiveTar(path.Dir(srcPath), path.Base(srcPath), path.Dir(destPath), path.Base(destPath), tarWriter)
+}
+
+// Based on code from https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/cp/cp.go
+func recursiveTar(srcBase, srcFile, destBase, destFile string, tw *tar.Writer) error {
+	srcPath := path.Join(srcBase, srcFile)
```

Comment:
Why can't we just use tar?

Reply:
here we write all the files into a pipe using `archive/tar` then stream to workload pod.